File: Samples/Layout/Tables/VB.NET/Tables.vb
﻿Imports BitMiracle.Docotic.Pdf.Layout

Namespace BitMiracle.Docotic.Pdf.Samples
    NotInheritable Class Tables
        Public Shared Sub Main()
            ' NOTE:
            ' Without a license, the library won't allow you to create or read PDF documents.
            ' To get a free time-limited license key, use the form on
            ' https://bitmiracle.com/pdf-library/download

            LicenseManager.AddLicenseData("PUT-LICENSE-HERE")

            Const PathToFile As String = "Tables.pdf"
            PdfDocumentBuilder.Create().Generate(PathToFile,
                Sub(doc)
                    doc.Pages(
                        Sub(page)
                            page.TextStyle(Function(t) t.Parent.FontSize(8)).
                                MarginVertical(50).
                                MarginHorizontal(10)

                            page.Content().Width(400).Table(
                                Sub(t)
                                    t.Columns(
                                        Sub(c)
                                            c.RelativeColumn(2)
                                            c.RelativeColumn(2)
                                            c.RelativeColumn(3)
                                        End Sub)

                                    t.Header(
                                        Sub(h)
                                            HeaderCell(h).Text("Project")
                                            HeaderCell(h).Text("License")
                                            HeaderCell(h).Text("Description")
                                        End Sub)

                                    For i As Integer = 0 To Products.Length - 1
                                        Dim alt As Boolean = i Mod 2 = 1
                                        Dim p As Product = Products(i)
                                        BodyCell(t, alt, p.Licenses.Length).Text(p.Name)

                                        For Each l As License In p.Licenses
                                            BodyCell(t, alt).Text(l.Name)
                                            BodyCell(t, alt).Text(l.Description)
                                        Next
                                    Next
                                End Sub)
                        End Sub)
                End Sub)

            Console.WriteLine($"The output is located in {Environment.CurrentDirectory}")
            Process.Start(New ProcessStartInfo(PathToFile) With {.UseShellExecute = True})
        End Sub

        Private Shared Function HeaderCell(t As TableCellContainer) As LayoutContainer
            Return t.
                Cell().
                Background(New PdfGrayColor(75)).
                Border(Function(b) b.Thickness(CellBorderThickness)).
                PaddingVertical(CellVerticalPadding).
                AlignCenter().
                AlignMiddle()
        End Function

        Private Shared Function BodyCell(t As Table, alt As Boolean, Optional rowSpan As Integer = 1) As LayoutContainer
            Return t.
                Cell(Sub(c) c.RowSpan(rowSpan)).
                Container(Function(c) If(alt, c.Background(New PdfGrayColor(90)), c)).
                Border(Function(b) b.Thickness(CellBorderThickness)).
                PaddingVertical(CellVerticalPadding).
                AlignCenter().
                AlignMiddle()
        End Function

        Private Const CellVerticalPadding = 10
        Private Const CellBorderThickness = 0.1

        Private Shared ReadOnly Products = New Product() {
            New Product("Docotic.Pdf", New License() {
                New License("Application License", "Desktop or mobile applications"),
                New License("Server License", "Server-based services"),
                New License("Ultimate License", "All your PDF needs"),
                New License("Utlimate+ License", "All your PDF needs + premium support")
            }),
            New Product("Jpeg2000.Net", New License() {New License("Ultimate License", "Any services or applications")}),
            New Product("LibTiff.Net", New License() {New License("Free", "")}),
            New Product("LibJpeg.Net", New License() {New License("Free", "")})
        }
    End Class

    NotInheritable Class Product
        Public Sub New(name As String, licenses As License())
            Me.Name = name
            Me.Licenses = licenses
        End Sub

        Public Name As String
        Public Licenses As License()
    End Class

    NotInheritable Class License
        Public Sub New(name As String, description As String)
            Me.Name = name
            Me.Description = description
        End Sub

        Public Name As String
        Public Description As String
    End Class
End Namespace

File: Samples/Layout/Tables/C#/Tables.cs
using System;
using System.Diagnostics;
using BitMiracle.Docotic.Pdf.Layout;

namespace BitMiracle.Docotic.Pdf.Samples
{
    static class Tables
    {
        static void Main()
        {
            // NOTE:
            // When used in trial mode, the library imposes some restrictions.
            // Please visit http://bitmiracle.com/pdf-library/trial-restrictions
            // for more information.

            const string PathToFile = "Tables.pdf";
            PdfDocumentBuilder.Create().Generate(PathToFile, doc =>
            {
                doc.Pages(page =>
                {
                    page.TextStyle(t => t.Parent.FontSize(8))
                        .MarginVertical(50)
                        .MarginHorizontal(10);

                    page.Content().Width(400).Table(t =>
                    {
                        t.Columns(c =>
                        {
                            c.RelativeColumn(2);
                            c.RelativeColumn(2);
                            c.RelativeColumn(3);
                        });

                        t.Header(h =>
                        {
                            HeaderCell(h).Text("Project");
                            HeaderCell(h).Text("License");
                            HeaderCell(h).Text("Description");
                        });

                        for (int i = 0; i < Products.Length; ++i)
                        {
                            bool alt = i % 2 == 1;
                            Product p = Products[i];
                            BodyCell(t, alt, p.Licenses.Length).Text(p.Name);
                            foreach (License l in p.Licenses)
                            {
                                BodyCell(t, alt).Text(l.Name);
                                BodyCell(t, alt).Text(l.Description);
                            }
                        }
                    });
                });
            });

            Console.WriteLine($"The output is located in {Environment.CurrentDirectory}");
            Process.Start(new ProcessStartInfo(PathToFile) { UseShellExecute = true });
        }

        private static LayoutContainer HeaderCell(TableCellContainer t)
        {
            return t.Cell()
                .Background(new PdfGrayColor(75))
                .Border(b => b.Thickness(CellBorderThickness))
                .PaddingVertical(CellVerticalPadding)
                .AlignCenter()
                .AlignMiddle();
        }

        private static LayoutContainer BodyCell(Table t, bool alt, int rowSpan = 1)
        {
            return t.Cell(c => c.RowSpan(rowSpan))
                .Container(c => alt ? c.Background(new PdfGrayColor(90)) : c)
                .Border(b => b.Thickness(CellBorderThickness))
                .PaddingVertical(CellVerticalPadding)
                .AlignCenter()
                .AlignMiddle();
        }

        private const double CellVerticalPadding = 10;
        private const double CellBorderThickness = 0.1;

        private static readonly Product[] Products = new Product[]
        {
            new("Docotic.Pdf", new License[]
            {
                new("Application License", "Desktop or mobile applications"),
                new("Server License", "Server-based services"),
                new("Ultimate License", "All your PDF needs"),
                new("Utlimate+ License", "All your PDF needs + premium support"),
            }),
            new("Jpeg2000.Net", new License[] { new("Ultimate License", "Any services or applications") }),
            new("LibTiff.Net", new License[] { new("Free", "") }),
            new("LibJpeg.Net", new License[] { new("Free", "") }),
        };
    }

    record Product(string Name, License[] Licenses);
    record License(string Name, string Description);
}